src/object/srv_csum.c

```diff
@@ -0,0 +1,555 @@
+ * bsgls in \ctx
+ */
+static int
+cc_add_csums_for_recx(struct csum_context *ctx, daos_recx_t *recx,
+		      daos_csum_buf_t *dcb)
+{
+	uint32_t		chunksize;
+	size_t			rec_size;
+	uint32_t		chunk_nr;
+	uint32_t		c;
+	int			rc = 0;
+
+	chunksize = daos_csummer_get_chunksize(ctx->cc_csummer);
+	rec_size = ctx->cc_rec_len;
+	chunk_nr = daos_recx_calc_chunks(*recx, rec_size, chunksize);
+
+	/** Because the biovs are acquired by searching for the recx, the first
+	 * selected/requested record of a biov will be the recx index
+	 */
+	ctx->cc_ext_start = recx->rx_idx;
+
+	cc_set_iov2ranges(ctx, cc2iov(ctx));
+
+	for (c = 0; c < chunk_nr; c++) { /** for each chunk/checksum */
+		ctx->cc_recx_chunk = csum_recx_chunkidx2range(recx, rec_size,
+							      chunksize, c);
+		ctx->cc_chunk = csum_chunkrange(chunksize / ctx->cc_rec_len,
+						c);
+		ctx->cc_chunk_bytes_left = ctx->cc_recx_chunk.dcr_nr *
+					   rec_size;
+
+		ctx->cc_csum_started = false;
+		cc_set_chunk2ranges(ctx);
+
+		/** need to loop until chunk bytes are consumed because more
+		 * than 1 extent might contribute to it
+		 */
+		while (ctx->cc_chunk_bytes_left > 0) {
+			/** All out of data. Just return because request may
+			 * be larger than previously written data
+			 */
+			if (ctx->cc_bsgl_idx->iov_idx >=
+			    ctx->cc_bsgl->bs_nr_out)
+				return 0;
+
+			rc = cc_add_csum(ctx, dcb, c);
+			if (rc != 0)
+				return rc;
+		}
+		if (ctx->cc_csum_started)
+			daos_csummer_finish(ctx->cc_csummer);
+
+		rc = cc_verify_orig_extents(ctx);
+		if (rc != 0)
+			return rc;
+		ctx->cc_to_verify_nr = 0;
+	}
+	cc_insert_remembered_csums(ctx);
+
+	return rc;
+}
+
+static uint64_t
+cc2biov_dcbs_nr(struct csum_context *ctx)
+{
+	return ctx->cc_biov_csum_idx;
+}
+
+int
+ds_csum_add2iod(daos_iod_t *iod, struct daos_csummer *csummer,
+		struct bio_sglist *bsgl, daos_csum_buf_t *biov_dcbs,
+		size_t *biov_dcbs_used)
+{
+	struct csum_context	ctx = {0};
+	struct daos_sgl_idx	bsgl_idx = {0};
+	int			rc = 0;
+	uint32_t		i;
+
+	if (!(daos_csummer_initialized(csummer) && iod->iod_recxs && bsgl))
+		return 0;
+
+	if (!csum_iod_is_supported(csummer->dcs_chunk_size, iod))
+		return 0;
+
+	/** setup the context */
+	ctx.cc_csummer = csummer;
+	ctx.cc_bsgl_idx = &bsgl_idx;
+	ctx.cc_rec_len = iod->iod_size;
+	ctx.cc_chunksize = csummer->dcs_chunk_size;
+	ctx.cc_bsgl = bsgl;
+	ctx.cc_biov_dcbs = biov_dcbs;
+	ctx.cc_to_verify = ctx.cc_to_verify_embedded;
+	ctx.cc_to_verify_size = TO_VERIFY_EMBEDDED_NR;
+
+	/** for each extent/checksum buf */
+	for (i = 0; i < iod->iod_nr && rc == 0; i++) {
+		daos_recx_t		*recx = &iod->iod_recxs[i];
+		daos_csum_buf_t		*dcb = &iod->iod_csums[i];
+
+		if (ctx.cc_rec_len > 0 && dcb_is_valid(dcb))
+			rc = cc_add_csums_for_recx(&ctx, recx, dcb);
+	}
+
+	/** return the cound of biov csums used. */
+	if (biov_dcbs_used != NULL)
+		*biov_dcbs_used = cc2biov_dcbs_nr(&ctx);
+
+	return rc;
+}
+
+bool
+ext_needs_new_csum(struct daos_csum_range *raw_ext,
```

Comment:
ds_csum_recomp_ext or something like this.

Reply:
renamed to ds_csum_calc_needed.